Psychtoolbox/PsychHardware/EyelinkToolbox/EyelinkBasic/EyelinkDoDriftCorrection.m

```diff
@@ -17,33 +17,97 @@
-	error = EyelinkDoDriftCorrect(el, x, y, draw, allowsetup);
-	
-	if error==el.TERMINATE_KEY
-		%result=el.ABORT_EXPT;
-		success=0;
-		return;
-	end;
-	% repeat if ESC was pressed to access Setup menu
-	if(error~=el.ESC_KEY) break; end
+while 1
+    if Eyelink('IsConnected')==el.notconnected   % Check link often so we don't lock up if tracker lost
+        success=0;
+        return;
+    end
+    
+    if ~isempty(el.callback) % if we have a callback set, we call it.
+        global eyelinkanimationtarget;
+        if isempty(eyelinkanimationtarget)
+            initmoviestruct();
+        end
+
+        global inDoTrackerSetup;
+        if isempty(inDoTrackerSetup)
+            inDoTrackerSetup = false;
+        end
+
+        global inDoDriftCorrection;
+        if isempty(inDoDriftCorrection)
+            inDoDriftCorrection = true;
+        end
+
+        if strcmpi(el.calTargetType, 'video') && ~eyelinkanimationtarget.init
+            loadanimationmovie(el);
+        end
+        result = Eyelink('DriftCorrStart', x, y, 1, draw, allowsetup);
+
+    else
+        % else we continue with the old version
+        result = EyelinkLegacyDoDriftCorrect(el, x, y, draw, allowsetup);
+    end
+    
+    if result==el.TERMINATE_KEY
+        success=0;
+        return;
+    end
+    
+    % repeat if ESC was pressed to access Setup menu
+    if(result~=el.ESC_KEY)
+        break;
+    end
+    
 end % while
+
+% fprintf('~isempty(el.callback): %d, el.calTargetType: %s, ~inDoTrackerSetup: %d, eyelinkanimationtarget.init: %d\n', ~isempty(el.callback), el.calTargetType, ~inDoTrackerSetup, eyelinkanimationtarget.init)
+if ~isempty(el.callback) && strcmpi(el.calTargetType, 'video') && ~inDoTrackerSetup && eyelinkanimationtarget.init
+    cleanupmovie(el);
+    inDoDriftCorrection = false;
+end
+
+return
+
+    function initmoviestruct()
+        eyelinkanimationtarget.init = false;
+        eyelinkanimationtarget.movie = [];
+        eyelinkanimationtarget.movieduration = [];
+        eyelinkanimationtarget.fps = [];
+        eyelinkanimationtarget.imgw = [];
+        eyelinkanimationtarget.imgh = [];
+        eyelinkanimationtarget.calxy =[];
+    end
+
+    function loadanimationmovie(el)
+        [movie, movieduration, fps, imgw, imgh] = Screen('OpenMovie',  el.window, el.calAnimationTargetFilename, el.calAnimationOpenAsync, el.calAnimationOpenPreloadSecs, el.calAnimationOpenSpecialFlags1);
+        eyelinkanimationtarget.init = true;
+        eyelinkanimationtarget.movie = movie;
+        eyelinkanimationtarget.movieduration = movieduration;
+        eyelinkanimationtarget.fps = fps;
+        eyelinkanimationtarget.imgw = imgw;
+        eyelinkanimationtarget.imgh = imgh;
+        Screen('SetMovieTimeIndex', eyelinkanimationtarget.movie, 0, el.calAnimationSetIndexIsFrames);
+    end
+
+    function cleanupmovie(el)
+        texkill = Screen('GetMovieImage', el.window, eyelinkanimationtarget.movie, el.calAnimationWaitTexClose);
```

Comment:
There was at one point early 2017 some buggy behaviour such that playing a video through to completion would cause graphics troubles thereafter. Can't recall if that was true freezing, or just hanging/spinning on gstreamer side. This was on Mac OS (not likely surprising) with El Capitan, a few PTB versions around 3.0.12 & 3.0.13, can't recall gstreamer version but matched around that time. Can't find what info I had at the time to resolve the problem, but I believe the trouble was something to do with video sink going stale, maybe it was a residual frame or junk in the buffer serving the frames. problematic behaviour would then present when interacting with the movie handle in any way (certainly rewinding and attempting to replay from start, but 99% certain even CloseMovie would lock up) unless GetMovieImage was first used to revive the stale state of interacting with gstreamer. perhaps this issue has been resolved whether the behaviour was related to implementation of gstreamer or gstreamer itself, etc. but left this 'texkill' convention in the code for posterity of the buggy behaviour in context it resolved at that time.

Reply:
Hm. I just reviewed our code: There haven't been any changes to the relevant code which could explain the appearance of such a bug around 2017, so it was almost certainly a GStreamer bug or problem on macOS, which likely has been fixed since then.

At least I don't think that I have ever observed the behavior you describe. Might be worth a retest and dropping this code. I'm not necessarily against keeping it if you insist, but it is definitely totally non-sensical and would be better to fix properly in Psychtoolbox than working around it this way, if the problem can still be reproduced.

Reply:
understood this can likely be done away with. Wish to leave it in for now, and save revisions to this part of the code -- that is distributed b/w DoCameraSetup, DoDriftCorrection, and the callback handlers /w shared structs, globals, etc. -- until a later time to include more thorough testing of the changes.